server/instance_cloud_oauth.go

```diff
@@ -0,0 +1,209 @@
+	tokenSource := oauth2Conf.TokenSource(ctx, connection.OAuth2Token)
+	client := oauth2.NewClient(ctx, tokenSource)
+
+	// Get a new token if Access Token has expired
+	currentToken := connection.OAuth2Token
+	updatedToken, err := tokenSource.Token()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error getting token from token source")
+	}
+
+	if updatedToken.RefreshToken != currentToken.RefreshToken {
+		connection.OAuth2Token = updatedToken
+
+		// Store this new access token & refresh token to get a new access token in future when it has expired
+		err = ci.Plugin.userStore.StoreConnection(ci.Common().InstanceID, connection.MattermostUserID, connection)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// TODO Get resource ID if not in the KV Store?
+	jiraID, err := ci.getJiraCloudResourceID(*client)
+	ci.JiraResourceID = jiraID
+	if err != nil {
+		return nil, nil, err
+	}
+
+	jiraClient, err := jira.NewClient(client, ci.GetURL())
+	return jiraClient, client, err
+}
+
+func (ci *cloudOAuthInstance) GetDisplayDetails() map[string]string {
+	return map[string]string{
+		"Jira Cloud Mattermost Key": ci.MattermostKey,
+	}
+}
+
+func (ci *cloudOAuthInstance) GetUserConnectURL(mattermostUserID string) (string, *http.Cookie, error) {
+	oauthConf := ci.GetOAuthConfig()
+	state := fmt.Sprintf("%s_%s", model.NewId()[0:15], mattermostUserID)
+	url := oauthConf.AuthCodeURL(
+		state,
+		oauth2.SetAuthURLParam("audience", "api.atlassian.com"),
+		oauth2.SetAuthURLParam("state", state),
+		oauth2.SetAuthURLParam("response_type", "code"),
+		oauth2.SetAuthURLParam("prompt", "consent"),
+	)
+	if err := ci.Plugin.otsStore.StoreOneTimeSecret(mattermostUserID, state); err != nil {
+		return "", nil, err
+	}
+	return url, nil, nil
+}
+
+func (ci *cloudOAuthInstance) GetOAuthConfig() *oauth2.Config {
+	return &oauth2.Config{
+		ClientID:     ci.JiraClientID,
+		ClientSecret: ci.JiraClientSecret,
+		Scopes:       strings.Split(JiraScopesOffline, ","),
+		RedirectURL:  fmt.Sprintf("%s%s", ci.Plugin.GetPluginURL(), instancePath(routeOAuth2Complete, ci.InstanceID)),
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  "https://auth.atlassian.com/authorize",
+			TokenURL: "https://auth.atlassian.com/oauth/token",
+		},
+	}
+}
+
+func (ci *cloudOAuthInstance) GetURL() string {
+	return "https://api.atlassian.com/ex/jira/" + ci.JiraResourceID
+}
+
+func (ci *cloudOAuthInstance) GetJiraBaseURL() string {
+	return ci.JiraBaseURL
+}
+
+func (ci *cloudOAuthInstance) GetManageAppsURL() string {
+	// TODO
+	return fmt.Sprintf("%s/plugins/servlet/applinks/listApplicationLinks", ci.GetURL())
+}
+
+func (ci *cloudOAuthInstance) GetManageWebhooksURL() string {
+	// TODO
+	return fmt.Sprintf("%s/plugins/servlet/webhooks", ci.GetURL())
+}
+
+func (ci *cloudOAuthInstance) GetMattermostKey() string {
+	return ci.MattermostKey
+}
+
+func (ci *cloudOAuthInstance) getJiraCloudResourceID(client http.Client) (string, error) {
+	request, err := http.NewRequest(
```

Comment:
We dont usually call any API directly see if they have functions like call or callJson and use them if possible ?

Reply:
There is no such method